### ISSUE TITLE
Rule proposal: Preferring .find(...) over .filter(...)[0]

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#6886]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

I've seen occasional code authors write code with a `.filter` returning element `[0]`, perhaps not knowing that `.find` does the same thing. Perhaps we could write a rule that flags for this, and uses the type system to know whether the collection being `.filter`ed is an array/array-like?

### Fail Cases

```typescript
const values = ['a', 'b ', 'c'];

const result = values.filter(value => value.includes(' '))[0];
```

### Pass Cases

```typescript
const values = ['a', 'b ', 'c'];

const result = values.find(value => value.includes(' '));
```

**Comment by armano2** [comment]
> My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.

https://github.com/WebbyLab/eslint-plugin-more/blob/master/docs/no-filter-instead-of-find.md

**Comment by jfmengels** [comment]
I guess there should ideally also be a check to see if the filter function does a side-effect, in which case the change will affect the current behavior.

**Comment by bradzacher** [comment]
> I guess there should ideally also be a check to see if the filter function does a side-effect

We can do that once we solve the halting problem 😄.
As long as it's not autofixing the code then reporting on it isn't bad.
Though TBH a side-effect in a `.filter` is filthy code and I want to break it to teach that person a lesson.

**Comment by jfmengels** [comment]
> We can do that once we solve the halting problem 😄.

Good luck, I'm looking forward to the results 😄 

> Though TBH a side-effect in a .filter is filthy code and I want to break it to teach that person a lesson.

I agree it is bad because it is unexpected. But I don't know about actively wanting to break it 😅 
I do agree that not having an autofix would solve the problem. A suggestion would be possible (with a warning about side-effects ideally?), but autofix would be bad because of how ESLint applies all fixes without a chance to review them.

Apart from that, I agree with the proposed rule (it's something I'd like for Elm as well but `find` is not a standard function unfortunately).

**Comment by domdomegg** [comment]
Is this already covered by https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-array-find.md?

**Comment by JoshuaKGoldberg** [comment]
We have existing precedent with https://typescript-eslint.io/rules/prefer-includes.

**Comment by JoshuaKGoldberg** [comment]
Oh and fun fact, https://sourcegraph.com/search?q=context:global+/filter.*%5C%29%5C%5B0%5C%5D/&patternType=standard&sm=1&groupBy=repo shows quite a few examples of people doing the `.filter(...)[0]` approach in the wild. Nifty!

**Comment by Josh-Cena** [comment]
I have a related request: I think you should prefer `some` over either `filter[0]` or `find` if the return value is only used in a Boolean context!

**Comment by JoshuaKGoldberg** [comment]
Whoo that's a complex one 😄. I don't know if it'd be easily feasible to get into this rule... maybe a good followup? Not sure :think

**Comment by kirkwaiblinger** [comment]
> I have a related request: I think you should prefer `some` over either `filter[0]` or `find` if the return value is only used in a Boolean context!

I wonder whether this ought to just be a whole separate rule. I could certainly try to include it as an option for the fix for .filter(...)[0], but then existing code like 
```
if ([1, 2, 3].find(x > 1)) {
}
```
would not get flagged. So maybe there should just be another rule entirely to prefer .some over .find in boolean contexts. I guess the drawback would be that if this rule just rewrites .filter(...)[0] -> .find(...) no matter what, then that will mean a second pass will be necessary to additionally fix the new rule.

Overall it seems like that rule would have quite a bit more complexity than makes sense just to add as a suggestion for this rule, since we'd have to start asking what we want done with 
```
if ([-1, 0, 1].find(x => x === 0)) {
    // doesn't execute; not the same as [-1, 0 , 1].some(x => x === 0)
}
```

Thoughts?
